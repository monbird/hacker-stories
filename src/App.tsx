import React from 'react';
import axios from 'axios';
import styles from './App.module.css';
import cs from 'classnames'; /* A simple javascript utility for conditionally joining classNames together */
import { ReactComponent as Check } from './check-solid.svg';
import { ReactComponent as Search } from './search-solid.svg';

import logo from './logo.svg';

const API_ENDPOINT = 'https://hn.algolia.com/api/v1/search?query=';

const useSemiPersistentState = (
    key: string,
    initialState: string
): [string, (newValue: string) => void] => {
    const isMounted = React.useRef(false);

    const [value, setValue] = React.useState(
        localStorage.getItem(key) || initialState
    );

    React.useEffect(() => {
        if (!isMounted.current) {
            // a way to prevent calling simple function in else statement for a component when rendering for the first time; useRef Hook keeps its ref.current property intact over re-renders
            isMounted.current = true;
        } else {
            localStorage.setItem(key, value);
        }
    }, [value, key]);

    return [value, setValue];
};

type Story = {
    objectID: string;
    url: string;
    title: string;
    author: string;
    num_comments: number;
    points: number;
};

type Stories = Array<Story>;

type StoriesState = {
    data: Stories;
    isLoading: boolean;
    isError: boolean;
};

interface StoriesFetchInitAction {
    type: 'STORIES_FETCH_INIT';
}

interface StoriesFetchSuccessAction {
    type: 'STORIES_FETCH_SUCCESS';
    payload: Stories;
}

interface StoriesFetchFailureAction {
    type: 'STORIES_FETCH_FAILURE';
}

interface StoriesRemoveAction {
    type: 'REMOVE_STORY';
    payload: Story;
}

type StoriesAction =
    | StoriesFetchInitAction
    | StoriesFetchSuccessAction
    | StoriesFetchFailureAction
    | StoriesRemoveAction;

const storiesReducer = (state: StoriesState, action: StoriesAction) => {
    switch (action.type) {
        case 'STORIES_FETCH_INIT':
            return {
                ...state,
                isLoading: true,
                isError: false,
            };
        case 'STORIES_FETCH_SUCCESS':
            return {
                ...state,
                isLoading: false,
                isError: false,
                data: action.payload,
            };
        case 'STORIES_FETCH_FAILURE':
            return {
                ...state,
                isLoading: false,
                isError: true,
            };
        case 'REMOVE_STORY':
            return {
                ...state,
                data: state.data.filter(
                    (story) => action.payload.objectID !== story.objectID
                ),
            };
        default:
            throw new Error();
    }
};

// TODO: make it typescript friendly
// const getSumComments = (stories) => {
//     return stories.data.reduce(
//         (result, value) => result + value.num_comments,
//         0
//     );
// };

const App = () => {
    const [searchTerm, setSearchTerm] = useSemiPersistentState(
        'search',
        'React'
    );

    const [url, setUrl] = React.useState(`${API_ENDPOINT}${searchTerm}`);

    const [stories, dispatchStories] = React.useReducer(storiesReducer, {
        data: [],
        isLoading: false,
        isError: false,
    });

    const handleFetchStories = React.useCallback(async () => {
        // useCallback creates/returns new version of this callback handler when one of its dependancies change
        dispatchStories({
            type: 'STORIES_FETCH_INIT',
        });

        try {
            const result = await axios.get(url);
            dispatchStories({
                type: 'STORIES_FETCH_SUCCESS',
                payload: result.data.hits,
            });
        } catch {
            dispatchStories({ type: 'STORIES_FETCH_FAILURE' });
        }
    }, [url]); // <- dependancy

    React.useEffect(() => {
        handleFetchStories();
    }, [handleFetchStories]);

    const handleRemoveStory = React.useCallback((item: Story) => {
        dispatchStories({
            type: 'REMOVE_STORY',
            payload: item,
        });
    }, []);

    const handleSearchInput = (event: React.ChangeEvent<HTMLInputElement>) => {
        setSearchTerm(event.target.value);
    };

    const handleSearchSubmit = (event: React.FormEvent<HTMLFormElement>) => {
        setUrl(`${API_ENDPOINT}${searchTerm}`);
        event.preventDefault();
    };

    // TODO: make it typescript friendly
    // const sumComments = React.useMemo(() => getSumComments(stories), [stories]); // useMemo returns memoized value from calling the function; then will call it only when there will be change in dependancy - stories

    return (
        <div className={styles.container}>
            <h1 className={styles.headlinePrimary}>
                {/* My Hacker Stories with {sumComments} comments. */}
                My Hacker Stories
            </h1>

            <img src={logo} className={styles.logo} alt="logo" />

            <SearchForm
                searchTerm={searchTerm}
                onSearchInput={handleSearchInput}
                onSearchSubmit={handleSearchSubmit}
            />

            <div role="button" tabIndex={0}>
                Click me
            </div>

            {stories.isError && <p>Something went wrong...</p>}

            {stories.isLoading ? (
                <p>Loading...</p>
            ) : (
                <List list={stories.data} onRemoveItem={handleRemoveStory} />
            )}
        </div>
    );
};

type SearchFormProps = {
    searchTerm: string;
    onSearchInput: (event: React.ChangeEvent<HTMLInputElement>) => void;
    onSearchSubmit: (event: React.FormEvent<HTMLFormElement>) => void;
};

const SearchForm = ({
    searchTerm,
    onSearchInput,
    onSearchSubmit,
}: SearchFormProps) => (
    <form onSubmit={onSearchSubmit} className={styles.searchForm}>
        <InputWithLabel
            id="search"
            value={searchTerm}
            isFocused
            onInputChange={onSearchInput}
        >
            Search:
            {/* <strong>Search: </strong> */}
        </InputWithLabel>
        <button
            type="submit"
            disabled={!searchTerm}
            className={cs(styles.button, styles.buttonLarge)}
            // className={cs(styles.button, { [styles.buttonLarge]: true })}    alternative
        >
            <Search height="16px" width="16px" />
            Submit
        </button>
    </form>
);

type InputWithLabelProps = {
    id: string;
    value: string;
    type?: string;
    onInputChange: (event: React.ChangeEvent<HTMLInputElement>) => void;
    isFocused?: boolean;
    children: React.ReactNode;
};

const InputWithLabel = ({
    id,
    value,
    type = 'text',
    onInputChange,
    isFocused,
    children, // use the children prop to access/render everything that has been passed down from above between the parent's element tags
}: InputWithLabelProps) => {
    const inputRef = React.useRef<HTMLInputElement>(null!);

    React.useEffect(() => {
        if (isFocused && inputRef.current) {
            inputRef.current.focus();
        }
    }, [isFocused]);

    return (
        <>
            <label htmlFor={id} className={styles.label}>
                {children}
            </label>
            &nbsp;
            <input
                ref={inputRef}
                id={id}
                type={type}
                value={value}
                autoFocus={isFocused}
                onChange={onInputChange}
                className={styles.input}
            />
        </>
    );
};

type ListProps = {
    list: Stories;
    onRemoveItem: (item: Story) => void;
};

const List = ({ list, onRemoveItem }: ListProps) => (
    <>
        {list.map((item) => (
            <Item key={item.objectID} item={item} onRemoveItem={onRemoveItem} />
        ))}
    </>
);

// const List = React.memo(({ list, onRemoveItem }) => // memo makes equality check for the props i.e component will not re-render if there was no change in props (list & onRemoveItem)
//     list.map((item) => (
//         <Item key={item.objectID} item={item} onRemoveItem={onRemoveItem} />
//     )));

type ItemProps = {
    item: Story;
    onRemoveItem: (item: Story) => void;
};

const Item = ({ item, onRemoveItem }: ItemProps) => (
    <div className={styles.item}>
        <span style={{ width: '40%' }}>
            <a href={item.url}>{item.title}</a>
        </span>
        <span style={{ width: '30%' }}>{item.author}</span>
        <span style={{ width: '10%' }}>{item.num_comments}</span>
        <span style={{ width: '10%' }}>{item.points}</span>
        <span style={{ width: '10%' }}>
            <button
                type="button"
                onClick={() => onRemoveItem(item)}
                className={`${styles.button} ${styles.buttonSmall}`}
            >
                <Check height="18px" width="18px" />
                Dismiss
            </button>
        </span>
    </div>
);

export default App;
export { storiesReducer, SearchForm, InputWithLabel, List, Item };
