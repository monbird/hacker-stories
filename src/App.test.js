import {
    render,
    screen,
    fireEvent,
    act,
    waitFor,
} from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import axios from 'axios';
import App, {
    storiesReducer,
    SearchForm,
    // InputWithLabel,
    // List,
    Item,
} from './App';

const storyOne = {
    title: 'React',
    url: 'https://reactjs.org/',
    author: 'Jordan Walke',
    num_comments: 3,
    points: 4,
    objectID: 0,
};

const storyTwo = {
    title: 'Redux',
    url: 'https://redux.js.org/',
    author: 'Dan Abramov, Andrew Clark',
    num_comments: 2,
    points: 5,
    objectID: 1,
};

const stories = [storyOne, storyTwo];

jest.mock('axios');

describe('storiesReducer', () => {
    test('removes a story from all stories', () => {
        // Setup
        const action = { type: 'REMOVE_STORY', payload: storyOne };
        const state = { data: stories, isLoading: false, isError: false };

        const expectedState = {
            data: [storyTwo],
            isLoading: false,
            isError: false,
        };

        // Exercise
        const newState = storiesReducer(state, action);

        // Verify
        expect(newState).toStrictEqual(expectedState);
    });

    test('stories fetching initialisation', () => {
        // Setup
        const action = { type: 'STORIES_FETCH_INIT' };
        const state = { data: [], isLoading: false, isError: false };

        const expectedState = {
            data: [],
            isLoading: true,
            isError: false,
        };

        // Exercise
        const newState = storiesReducer(state, action);

        // Verify
        expect(newState).toStrictEqual(expectedState);
    });

    test('stories fetching success', () => {
        // Setup
        const action = { type: 'STORIES_FETCH_SUCCESS', payload: stories };
        const state = { data: [], isLoading: false, isError: false };

        const expectedState = {
            data: stories,
            isLoading: false,
            isError: false,
        };

        // Exercise
        const newState = storiesReducer(state, action);

        // Verify
        expect(newState).toStrictEqual(expectedState);
    });

    test('stories fetching failure', () => {
        // Setup
        const action = { type: 'STORIES_FETCH_FAILURE' };
        const state = { data: [], isLoading: false, isError: false };

        const expectedState = {
            data: [],
            isLoading: false,
            isError: true,
        };

        // Exercise
        const newState = storiesReducer(state, action);

        // Verify
        expect(newState).toStrictEqual(expectedState);
    });
});

describe('Item', () => {
    test('renders all properties', () => {
        render(<Item item={storyOne} />);
        // screen.debug();
        expect(screen.getByText('Jordan Walke')).toBeInTheDocument();
        expect(screen.getByText('React')).toHaveAttribute(
            'href',
            'https://reactjs.org/'
        );
    });

    test('renders a clickable dismiss button', () => {
        render(<Item item={storyOne} />);

        expect(screen.getByRole('button')).toBeInTheDocument();
    });

    test('clicking the dismiss button calls the callback handler', () => {
        const handleRemoveItem = jest.fn();
        // const handleRemoveItem = jest.fn(() => {
        //     return 'hello';
        // });

        render(<Item item={storyOne} onRemoveItem={handleRemoveItem} />);

        fireEvent.click(screen.getByRole('button'));
        // userEvent.click(screen.getByRole('button'));

        expect(handleRemoveItem).toHaveBeenCalledTimes(1);
        // expect(handleRemoveItem()).toBe('hello');
    });

    test('renders snapshot', () => {
        const { container } = render(<Item item={storyOne} />);
        expect(container.firstChild).toMatchSnapshot();
    });
});

describe('SearchForm', () => {
    const SearchFormProps = {
        searchTerm: 'React',
        onSearchInput: jest.fn(),
        onSearchSubmit: jest.fn(),
    };

    test('renders the input field with its value', () => {
        render(<SearchForm {...SearchFormProps} />);
        // screen.debug();
        expect(screen.getByDisplayValue('React')).toBeInTheDocument();
    });

    test('renders the correct label', () => {
        render(<SearchForm {...SearchFormProps} />);
        expect(screen.getByLabelText(/Search/)).toBeInTheDocument();
    });

    test('calls onSearchInput on input field change', () => {
        render(<SearchForm {...SearchFormProps} />);

        // fireEvent.change(screen.getByDisplayValue('React'), {
        //     target: { value: 'Redux' },
        // });

        // userEvent.paste(screen.getByDisplayValue('React'), 'Redux');
        // expect(SearchFormProps.onSearchInput).toHaveBeenCalledTimes(1);

        userEvent.type(screen.getByDisplayValue('React'), 'Redux');
        expect(SearchFormProps.onSearchInput).toHaveBeenCalledTimes(5);
    });

    test('calls onSearchSubmit on button submit click', () => {
        render(<SearchForm {...SearchFormProps} />);
        screen.debug();
        fireEvent.submit(screen.getByRole('button'));

        expect(SearchFormProps.onSearchSubmit).toHaveBeenCalledTimes(1);
    });

    test('renders snapshot', () => {
        const { container } = render(<SearchForm {...SearchFormProps} />);
        expect(container.firstChild).toMatchSnapshot();
    });
});

describe('App', () => {
    test('succeeds fetching data', async () => {
        const promise = Promise.resolve({ data: { hits: stories } });
        axios.get.mockImplementationOnce(() => promise);

        render(<App />);
        // screen.debug();
        expect(screen.queryByText(/loading/i)).toBeInTheDocument();

        // await act(() => promise);
        await waitFor(() => promise);
        // screen.debug();
        expect(screen.queryByText(/loading/i)).toBeNull();

        expect(screen.getByText('React')).toBeInTheDocument();
        expect(screen.getByText('Redux')).toBeInTheDocument();
        expect(screen.getAllByText('Dismiss').length).toBe(2);
    });

    test('fails fetching data', async () => {
        const promise = Promise.reject();
        axios.get.mockImplementationOnce(() => promise);

        render(<App />);

        expect(screen.queryByText(/Loading/)).toBeInTheDocument();

        try {
            await waitFor(() => promise);
        } catch (error) {
            // eslint-disable-next-line jest/no-conditional-expect
            expect(screen.queryByText(/Loading/)).toBeNull();
            // eslint-disable-next-line jest/no-conditional-expect
            expect(screen.queryByText(/went wrong/)).toBeInTheDocument();
        }
    });

    test('removes a story', async () => {
        const promise = Promise.resolve({ data: { hits: stories } });
        axios.get.mockImplementationOnce(() => promise);

        render(<App />);

        await waitFor(() => promise);

        expect(screen.getAllByText('Dismiss').length).toBe(2);
        expect(screen.getByText('Jordan Walke')).toBeInTheDocument();

        userEvent.click(screen.getAllByText('Dismiss')[0]);

        expect(screen.getAllByText('Dismiss').length).toBe(1);
        expect(screen.queryByText('Jordan Walke')).toBeNull();
    });

    test('searches for specific stories', async () => {
        const reactPromise = Promise.resolve({ data: { hits: stories } });

        const anotherStory = {
            title: 'JavaScript',
            url: 'https://en.wikipedia.org/wiki/JavaScript',
            author: 'Brendan Eich',
            num_comments: 15,
            points: 10,
            objectID: 3,
        };

        const javascriptPromise = Promise.resolve({
            data: { hits: [anotherStory] },
        });

        axios.get.mockImplementation((url) => {
            if (url.includes('React')) {
                return reactPromise;
            }
            if (url.includes('JavaScript')) {
                return javascriptPromise;
            }

            throw Error();
        });

        // Initial render
        render(<App />);

        // 1. First data fetching
        await waitFor(() => reactPromise);

        expect(screen.queryByDisplayValue('React')).toBeInTheDocument();
        expect(screen.queryByDisplayValue('JavaScript')).toBeNull();
        expect(screen.queryByText('Jordan Walke')).toBeInTheDocument();
        expect(
            screen.queryByText('Dan Abramov, Andrew Clark')
        ).toBeInTheDocument();
        expect(screen.queryByText('Brendan Eich')).toBeNull();

        // 2. User interaction: changing input field value in Search component
        // fireEvent.change(screen.queryByDisplayValue('React'), {
        //     target: { value: 'JavaScript' },
        // });
        const input = screen.queryByDisplayValue('React');
        userEvent.clear(input);
        userEvent.type(input, 'JavaScript');

        expect(screen.queryByDisplayValue('React')).toBeNull();
        expect(screen.queryByDisplayValue('JavaScript')).toBeInTheDocument();

        // 3. Submitting the form
        // fireEvent.submit(screen.queryByText('Submit'));
        userEvent.click(screen.queryByText('Submit'));

        // 4. Second data fetching: retriving new data from the API
        await waitFor(() => javascriptPromise);

        expect(screen.queryByText('Jordan Walke')).toBeNull();
        expect(screen.queryByText('Dan Abramov, Andrew Clark')).toBeNull();
        expect(screen.queryByText('Brendan Eich')).toBeInTheDocument();
    });
});
