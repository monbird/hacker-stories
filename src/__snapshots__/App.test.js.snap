// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Item renders snapshot 1`] = `
<div
  class="item"
>
  <span
    style="width: 40%;"
  >
    <a
      href="https://reactjs.org/"
    >
      React
    </a>
  </span>
  <span
    style="width: 30%;"
  >
    Jordan Walke
  </span>
  <span
    style="width: 10%;"
  >
    3
  </span>
  <span
    style="width: 10%;"
  >
    4
  </span>
  <span
    style="width: 10%;"
  >
    <button
      class="button buttonSmall"
      type="button"
    >
      <svg
        height="18px"
        width="18px"
      >
        check-solid.svg
      </svg>
      Dismiss
    </button>
  </span>
</div>
`;

exports[`SearchForm renders snapshot 1`] = `
<form
  class="searchForm"
>
  <label
    class="label"
    for="search"
  >
    Search:
  </label>
  Â 
  <input
    class="input"
    id="search"
    type="text"
    value="React"
  />
  <button
    class="button buttonLarge"
    type="submit"
  >
    <svg
      height="16px"
      width="16px"
    >
      search-solid.svg
    </svg>
    Submit
  </button>
</form>
`;
